src/app_ui/address.rs

```diff
@@ -15,29 +15,28 @@
  *  limitations under the License.
  *****************************************************************************/
 
-use crate::utils::{concatenate, to_hex_all_caps};
 use crate::AppSW;
-use core::str::from_utf8;
+use core::str::{from_utf8, from_utf8_mut};
 use ledger_device_ui_sdk::bitmaps::{CROSSMARK, EYE, VALIDATE_14};
 use ledger_device_ui_sdk::ui::{Field, MultiFieldReview};
 
 // Display only the last 20 bytes of the address
 const DISPLAY_ADDR_BYTES_LEN: usize = 20;
 
 pub fn ui_display_pk(addr: &[u8]) -> Result<bool, AppSW> {
-    let addr_hex_str_buf = to_hex_all_caps(&addr[addr.len() - DISPLAY_ADDR_BYTES_LEN as usize..])
-        .map_err(|_| AppSW::AddrDisplayFail)?;
-    let addr_hex_str = from_utf8(&addr_hex_str_buf[..DISPLAY_ADDR_BYTES_LEN * 2])
-        .map_err(|_| AppSW::AddrDisplayFail)?;
-
-    let mut addr_hex_str_with_prefix_buf = [0u8; DISPLAY_ADDR_BYTES_LEN * 2 + 2];
-    concatenate(&["0x", &addr_hex_str], &mut addr_hex_str_with_prefix_buf);
-    let addr_hex_str_with_prefix =
-        from_utf8(&addr_hex_str_with_prefix_buf).map_err(|_| AppSW::AddrDisplayFail)?;
+    let mut addr_hex = [0u8; DISPLAY_ADDR_BYTES_LEN * 2 + 2];
+    addr_hex[..2].copy_from_slice("0x".as_bytes());
+    hex::encode_to_slice(
+        &addr[addr.len() - DISPLAY_ADDR_BYTES_LEN..],
+        &mut addr_hex[2..],
+    )
+    .unwrap();
```

Comment:
You should probably use `map_err` with the `AddrDisplayFail` `AppSW` and `?` to propagate errors @kingofpayne.

You can do that everywhere you use `unwrap` in this function (line 33 and line 34).

Reply:
Calls to `unwrap` are in those case not supposed to fail in any way. Calls to `unwrap` are to me legitimate as the contracts on the inputs are respected. This avoid forwarding errors unnecessarily and reduces code complexity.